### PR TITLE
[traffic-gen-in-vm] Add traffic-gen build system

### PR DESCRIPTION
**Pull request by orelmisan**
Add the required files and logic in order to build an push the traffic-gen containerDisk image.
This includes:
1. virt-builder scripts.
2. Makefile targets.
3. GitHub actions workflows to automatically build and push it.

In order to build the traffic-gen's containerDisk image:
```bash
make build-traffic-gen-container-disk
```

In order to push it:
```bash
push-traffic-gen-container-disk
```

~~Depends on PR #122, please review only the last eight commits.~~

### CODE REVIEW

**Comment by orelmisan** [comment]
[Change](https://github.com/kiagnose/kubevirt-dpdk-checkup/compare/af5a242dd102348854ce61de766316a79b92604c..98007c58fb21ffd296261c0aa0f66938f589d892): Added TRex installation.

**Comment by orelmisan** [comment]
The traffic-gen container disk image was manually tested with the following VMI spec:
```yaml
apiVersion: kubevirt.io/v1
kind: VirtualMachineInstance
metadata:
  annotations:
  name: traffic-gen
spec:
  domain:
    cpu:
      cores: 4
      dedicatedCpuPlacement: true
      model: host-model
      sockets: 1
      threads: 2
    devices:
      disks:
        - disk:
            bus: virtio
          name: rootdisk
        - disk:
            bus: virtio
          name: cloudinitdisk
      interfaces:
        - macAddress: '60:00:c7:7f:75:01'
          model: virtio
          name: nic-east
          pciAddress: '0000:06:00.0'
          sriov: {}
        - macAddress: '60:f3:6a:0e:8f:02'
          model: virtio
          name: nic-west
          pciAddress: '0000:07:00.0'
          sriov: {}
      networkInterfaceMultiqueue: true
      rng: {}
    memory:
      guest: 4Gi
      hugepages:
        pageSize: 1Gi
  networks:
    - multus:
        networkName: dpdk-sriovnetwork-ns-2
      name: nic-east
    - multus:
        networkName: dpdk-sriovnetwork-ns-2
      name: nic-west
  terminationGracePeriodSeconds: 0
  volumes:
    - containerDisk:
        image: 'quay.io/omisan/kubevirt-dpdk-checkup-traffic-gen:latest'
        imagePullPolicy: Always
      name: rootdisk
    - cloudInitNoCloud:
        userData: |-
          #cloud-config
          user: cloud-user
          password: 0tli-pxem-xknu
          chpasswd:
            expire: false
      name: cloudinitdisk
```

**Comment by orelmisan** [comment]
[Change](https://github.com/kiagnose/kubevirt-dpdk-checkup/compare/98007c58fb21ffd296261c0aa0f66938f589d892..22c7d2cb0e98f887abc72ecd226ee3ba183a1618): Rebase

**Comment by orelmisan** [comment]
[Change](https://github.com/kiagnose/kubevirt-dpdk-checkup/compare/22c7d2cb0e98f887abc72ecd226ee3ba183a1618..3475a2212740c591f9865c71aac2a3c09bd53c2c): Rebase

**Review by RamLavi — CHANGES_REQUESTED** [review]
looks good overall, added one comment

File: vms/traffic-gen/scripts/first-boot [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Copyright 2023 Red Hat, Inc.
+#
+
+driverctl set-override 0000:06:00.0 vfio-pci
```
[the remark this answers is not in the document]
Comment (orelmisan) [reply vms/traffic-gen/scripts/first-boot]: Done.

**Comment by orelmisan** [comment]
[Change](https://github.com/kiagnose/kubevirt-dpdk-checkup/compare/3475a2212740c591f9865c71aac2a3c09bd53c2c..2ddac9bac2db7bc90c75bbe8e04cf1a809d37839): Answered review [comment](https://github.com/kiagnose/kubevirt-dpdk-checkup/pull/126#discussion_r1259628429).

**Comment by RamLavi** [comment]
Shouldn't we keep the quay naming consistent?
DPDK VM : https://quay.io/repository/kiagnose/kubevirt-dpdk-checkup-vm
traffic-gen VM: https://quay.io/repository/kiagnose/kubevirt-dpdk-checkup-traffic-gen-vm
?

**Comment by orelmisan** [comment]
> Shouldn't we keep the quay naming consistent? DPDK VM : https://quay.io/repository/kiagnose/kubevirt-dpdk-checkup-vm ---> traffic-gen VM: https://quay.io/repository/kiagnose/kubevirt-dpdk-checkup-traffic-gen-vm ?

My intention was to use the existing registry for the traffic generator.
I think that creating another registry will put a maintenance burden on us.

**Comment by RamLavi** [comment]
> > Shouldn't we keep the quay naming consistent? DPDK VM : https://quay.io/repository/kiagnose/kubevirt-dpdk-checkup-vm ---> traffic-gen VM: https://quay.io/repository/kiagnose/kubevirt-dpdk-checkup-traffic-gen-vm ?
> 
> My intention was to use the existing registry for the traffic generator. I think that creating another registry will put a maintenance burden on us.

valid decision

File: vms/traffic-gen/scripts/build-vm-image [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+#
+# Copyright 2023 Red Hat, Inc.
+#
+
+export LIBGUESTFS_BACKEND=direct
+
+virt-builder centosstream-8 \
+  --format qcow2 \
+  --root-password password:redhat \
+  --install cloud-init,driverctl,tuned-profiles-cpu-partitioning,tar,python3,pciutils \
```
Comment (RamLavi) [inline vms/traffic-gen/scripts/build-vm-image:25]: aren't we missing a lot of packages? https://github.com/kiagnose/kubevirt-dpdk-checkup/blob/main/traffic-gen/Dockerfile#L7

**Comment by RamLavi** [comment]
Do you plan to add the trex scripts and template in another PR right? https://github.com/kiagnose/kubevirt-dpdk-checkup/blob/main/traffic-gen/Dockerfile#L34C1-L35C30

**Comment by RamLavi** [comment]
just a small note: This PR is installing trex on the VM but trex is not "alive" yet. Meaning it needs to run in the background. We can add it a daemon in another PR. wdyt?

**Comment by orelmisan** [comment]
> just a small note: This PR is installing trex on the VM but trex is not "alive" yet. Meaning it needs to run in the background. We can add it a daemon in another PR. wdyt?

We need to do a little design on how we plan to configure and run TRex.
There are several options to do so.

File: vms/traffic-gen/scripts/build-vm-image (hunk 2 above)
Comment (orelmisan) [reply vms/traffic-gen/scripts/build-vm-image:25]: Is there a package you saw that is missing after manually running TRex?

Comment (RamLavi) [reply vms/traffic-gen/scripts/build-vm-image:25]: I didn't run it yet. In order to run it one would need to add the trex scripts

**Comment by RamLavi** [comment]
IIUC `--firstboot` runs only once, and `--run` runs on each reboot. right?
If this is currect, then does it make sense if we run the customize-vm as `firstboot` (AFAIK nothing there needs to run at each reboot right?) and run the trex-configuration scripts (that are not in this PR yet) as `--run`?

**Review by RamLavi — APPROVED** [review]
/lgtm
/approve

**Pull request merged**